### PR TITLE
Add semantic conventions for GraphQL

### PR DESCRIPTION
**Pull request by laurit**
## Changes

Add semantic conventions for GraphQL span name and attributes.

Related issues #1670

### CODE REVIEW

File: specification/trace/semantic_conventions/instrumentation/graphql.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# Semantic conventions for GraphQL Server
+
+**Status**: [Experimental](../../../document-status.md)
+
+This document defines semantic conventions to apply when instrumenting the GraphQL Server. They map GraphQL query to
+attributes on a Span.
```
Comment (trask) [inline specification/trace/semantic_conventions/instrumentation/graphql.md:6]:
```suggestion
This document defines semantic conventions to apply when instrumenting the GraphQL Server. They map GraphQL queries to
attributes on a Span.
```

File: specification/trace/semantic_conventions/instrumentation/graphql.md [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+<!-- semconv graphql -->
+| Attribute  | Type | Description  | Examples  | Required |
+|---|---|---|---|---|
+| `graphql.graphql.operation.name` | string | The name of the query being executed. | `findBookById` | No |
+| `graphql.graphql.operation.type` | string | The name of the operation being executed. | `QUERY`; `MUTATION`; `SUBSCRIPTION` | No |
+| `graphql.graphql.source` | string | The GraphQL query being executed. [1] | `query findBookById { bookById(id: ?) { name } }` | No |
```
Comment (trask) [inline specification/trace/semantic_conventions/instrumentation/graphql.md:13]: `graphql.graphql.` -> `graphql.`

File: specification/trace/semantic_conventions/instrumentation/graphql.md (hunk 1 above)
Comment (laurit) [reply specification/trace/semantic_conventions/instrumentation/graphql.md:6]: thanks

File: specification/trace/semantic_conventions/instrumentation/graphql.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+This document defines semantic conventions to apply when instrumenting the GraphQL Server. They map GraphQL queries to
+attributes on a Span.
+
+<!-- semconv graphql -->
+| Attribute  | Type | Description  | Examples  | Required |
+|---|---|---|---|---|
+| `graphql.operation.name` | string | The name of the query being executed. | `findBookById` | No |
+| `graphql.operation.type` | string | The name of the operation being executed. | `QUERY`; `MUTATION`; `SUBSCRIPTION` | No |
+| `graphql.source` | string | The GraphQL query being executed. [1] | `query findBookById { bookById(id: ?) { name } }` | No |
```
[the remark this answers is not in the document]
Comment (laurit) [reply specification/trace/semantic_conventions/instrumentation/graphql.md]: I only used `graphql.source` because that is what was used in js and java (which copied from js), was hoping that someone would suggest to rename this.
As far as I can tell standard GraphQL server allows only one operation. Execution fails when input query contains multiple operations and operation name is not specified.

File: specification/trace/semantic_conventions/instrumentation/graphql.md [hunk 4]
```diff
@@ -0,0 +1,16 @@
+# Semantic conventions for GraphQL Server
+
+**Status**: [Experimental](../../../document-status.md)
+
+This document defines semantic conventions to apply when instrumenting the GraphQL Server. They map GraphQL queries to
+attributes on a Span.
+
```
Comment (mateuszrzeszutek) [inline specification/trace/semantic_conventions/instrumentation/graphql.md:7]: What about the span name? Can we specify `${graphql.operation.type} ${graphql.operation.name}` as the expected span name for graphQL spans?

Comment (trask) [reply specification/trace/semantic_conventions/instrumentation/graphql.md:7]: Also a note about span kind would be good, since I think both Java and JavaScript(?) are capturing these spans as `Internal`, which could be non-obvious to spec readers.

File: semantic_conventions/trace/instrumentation/graphql.yml [hunk 5]
```diff
@@ -0,0 +1,20 @@
+groups:
+  - id: graphql
+    prefix: graphql
+    brief: >
+      This document defines semantic conventions to apply when instrumenting the GraphQL Server. They map GraphQL queries
```
Comment (svrnm) [inline semantic_conventions/trace/instrumentation/graphql.yml:5]: >  when instrumenting the GraphQL Server.

Is this really only applicable to the server? what if someone talks to a non-instrumented graphql backend and wants to use the same semantics at the client side?

Comment (svrnm) [inline semantic_conventions/trace/instrumentation/graphql.yml:5]: >  when instrumenting the GraphQL Server.

Is this really only applicable to the server? what if someone talks to a non-instrumented graphql backend and wants to use the same semantics at the client side?

**Comment by svrnm** [comment]
Thanks for creating this @laurit !

File: semantic_conventions/trace/instrumentation/graphql.yml [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+      to attributes on a Span.
+    attributes:
+      - id: operation.name
```
Comment (yurishkuro) [inline semantic_conventions/trace/instrumentation/graphql.yml:8]: how is this related to the span name?

File: specification/trace/semantic_conventions/instrumentation/graphql.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (laurit) [reply specification/trace/semantic_conventions/instrumentation/graphql.md]: changed `graphql.query` -> `graphql.operation.body` and replaced `query` with `operation` in descriptions.

**Review by mateuszrzeszutek — APPROVED** [review]

File: specification/trace/semantic_conventions/instrumentation/graphql.md (hunk 4 above)
Comment (laurit) [reply specification/trace/semantic_conventions/instrumentation/graphql.md:14]: Added span kind and span name.

File: specification/trace/semantic_conventions/instrumentation/graphql.md [hunk 7]
```diff
@@ -0,0 +1,23 @@
+# Semantic conventions for GraphQL Server
+
+**Status**: [Experimental](../../../document-status.md)
+
+This document defines semantic conventions to apply when instrumenting the GraphQL Server. They map GraphQL operations
+to attributes on a Span.
+
+**Span kind:** MUST always be `INTERNAL`.
```
Comment (yurishkuro) [inline specification/trace/semantic_conventions/instrumentation/graphql.md:8]: Q: why is this a requirement to create a new span instead of annotating the existing client/server spans?

Comment (laurit) [reply specification/trace/semantic_conventions/instrumentation/graphql.md:8]: My understanding is that adding these attributes to an existing server span could make sense when you have built an api that consumes graphql. At least in the graphql java framework there is nothing that requires it to be used that way. In my opinion adding these attributes to an existing server span would not make sense if multiple graphql operations are executed during one request and when the same request performs other operations besides executing qraphql.

Comment (trask) [reply specification/trace/semantic_conventions/instrumentation/graphql.md:8]: > My understanding is that adding these attributes to an existing server span could make sense when you have built an api that consumes graphql

maybe relax the requirement to allow it to be either an INTERNAL or SERVER span?

Comment (svrnm) [reply specification/trace/semantic_conventions/instrumentation/graphql.md:8]: what's the purpose of this limitation? as per my comment above, it is possible that someone wants to use this semantic conventions in a client side request, especially when the downstream service is not instrumented/owned by a 3rd party

File: semantic_conventions/trace/instrumentation/graphql.yml (hunk 5 above)
Comment (svrnm) [reply semantic_conventions/trace/instrumentation/graphql.yml]: I would prefer `implementation` yes. By that there are no implicit assumptions where those semantics are used.

Comment (laurit) [reply semantic_conventions/trace/instrumentation/graphql.yml]: Replaced `server` with `implementation`

File: specification/trace/semantic_conventions/instrumentation/graphql.md (hunk 7 above)
Comment (laurit) [reply specification/trace/semantic_conventions/instrumentation/graphql.md]: Removed span kind.

File: specification/trace/semantic_conventions/instrumentation/graphql.md [hunk 8]
```diff
@@ -0,0 +1,21 @@
+# Semantic conventions for GraphQL Server
```
Comment (trask) [inline specification/trace/semantic_conventions/instrumentation/graphql.md:1]: based on discussion that this could be used for clients too

```suggestion
# Semantic conventions for GraphQL
```

**Review by trask — COMMENTED** [review]
does anyone have thoughts on dropping `operation` from the attribute names, e.g.

* `graphql.name`
* `graphql.type`
* `graphql.body`

**Review by trask — APPROVED** [review]

**Comment by svrnm** [comment]
> does anyone have thoughts on dropping `operation` from the attribute names, e.g.

I would keep it, there are other _things_ in graphql that might have a name, type or body (not an expert on that, but by quickly reading the [GraphQL](https://spec.graphql.org/October2021/) spec I saw that there are also things like type definitions, fragments, response. Also, their spec could extend to contain other objects worth collecting)

While those other things are not yet specified here, one might do it in the future and naming it `graphql.name|type|body` would be limiting that effort.

File: specification/trace/semantic_conventions/instrumentation/graphql.md [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+MAY be used as span name.
+
+<!-- semconv graphql -->
+| Attribute  | Type | Description  | Examples  | Required |
+|---|---|---|---|---|
+| `graphql.operation.name` | string | The name of the operation being executed. | `findBookById` | No |
+| `graphql.operation.type` | string | The type of the operation being executed. | `QUERY`; `MUTATION`; `SUBSCRIPTION` | No |
+| `graphql.operation.body` | string | The GraphQL query being executed. [1] | `query findBookById { bookById(id: ?) { name } }` | No |
+
+**[1]:** The value may be sanitized to exclude sensitive information.
```
Comment (jamesmoessis) [inline specification/trace/semantic_conventions/instrumentation/graphql.md:20]: In the currently available instrumentation, is there any sanitization? The operation body seems like something that instrumentation might not have enough context to sanitize.

Comment (laurit) [reply specification/trace/semantic_conventions/instrumentation/graphql.md:20]: There is sanitation in java instrumentation and as far as I can tell in js also.

**Comment by trask** [comment]
/unstale

@yurishkuro @svrnm @jamesmoessis any other thoughts on this?

**Review by svrnm — APPROVED** [review]

**Comment by svrnm** [comment]
@trask, thanks, I was under the assumption that this was closed already. I'm fine with it!

**Review by jamesmoessis — APPROVED** [review]

**Comment by trask** [comment]
@open-telemetry/technical-committee any thoughts on the bar for merging this? thx

File: specification/trace/semantic_conventions/instrumentation/graphql.md [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+The **span name** MUST be of the format `<graphql.operation.type> <graphql.operation.name>` provided that
+`graphql.operation.type` and `graphql.operation.name` are available. If `graphql.operation.name` is not available, the
+span SHOULD be named `<graphql.operation.type>`. When `<graphql.operation.type>` is not available, `GraphQL Operation`
+MAY be used as span name.
+
+<!-- semconv graphql -->
+| Attribute  | Type | Description  | Examples  | Required |
+|---|---|---|---|---|
+| `graphql.operation.name` | string | The name of the operation being executed. | `findBookById` | No |
+| `graphql.operation.type` | string | The type of the operation being executed. | `QUERY`; `MUTATION`; `SUBSCRIPTION` | No |
```
Comment (tigrannajaryan) [inline specification/trace/semantic_conventions/instrumentation/graphql.md:17]: Is type expected to be capitalized? Graphql spec uses lowercase. Is this a value from a finite set (enum of 3 possible values) or arbitrary values are possible?

**Comment by tigrannajaryan** [comment]
@spencerwilson @ericmustin @jord1e please review.

File: specification/trace/semantic_conventions/instrumentation/graphql.md (hunk 10 above)
Comment (tigrannajaryan) [reply specification/trace/semantic_conventions/instrumentation/graphql.md:17]: There is a way to limit the values to a finite enum, see for example https://github.com/open-telemetry/opentelemetry-specification/blob/031630c818c60666b27764a5b7a0e4ed435f55c4/semantic_conventions/trace/messaging.yaml#L102

Comment (tigrannajaryan) [reply specification/trace/semantic_conventions/instrumentation/graphql.md:17]: I think using the lowercase like Graphql spec does would be a bit more preferable.

Comment (laurit) [reply specification/trace/semantic_conventions/instrumentation/graphql.md]: Done

**Comment by jord1e** [comment]
Looks good, but shouldn't we add tracing for parsing and validation steps? And maybe for individual fields? - (GraphQL Java has the concept of [`TrivialDataFetcher`s](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/TrivialDataFetcher.java) which can be used as a marker interface for tracing exclusion, I don't know if other GraphQL implementations have an equivalent)

**Comment by laurit** [comment]
> Looks good, but shouldn't we add tracing for [_parsing_](https://spec.graphql.org/October2021/#sel-EANLDCAACC5Ev_f) and [_validation_](https://spec.graphql.org/October2021/#sec-Validating-Requests) steps? They can contribute significantly to simple (mainly parsing) and complex (mainly validation) operations.
> 
Not specifying parsing, validation and data fetchers was a deliberate choice. In java instrumentation we don't have instrumentations that would provide that level of detail, hence our GraphQL instrumentation also currently also generates only a single span that encompasses the whole operation. If requested by users we may add extra spans, possibly enabled by a configuration flag. Instrumentations in other languages might choose to handle this differently and generate more spans. This PR only aims to find some common ground by defining attributes that could be useful for all GraphQL instrumentations and the name of the top level span.

**Comment by laurit** [comment]
> * This Spec PR has no real details around Error handling, which I think is important. We have some handling for it here, it might make sense to formalize this, https://github.com/open-telemetry/opentelemetry-ruby/blob/46ef191f579d23ce51763a50c01ab6a43a233adf/instrumentation/graphql/lib/opentelemetry/instrumentation/graphql/tracers/graphql_tracer.rb#L32-L39
> 
As far as I can tell this only handles validation errors, but at least in the java implementation validation isn't the only source for errors. If people feel that we should define an event for errors I'd suggest to use a more generic name than `graphql.validation.error`. Currently both java and javascript implementation report such errors as exceptions although they really aren't exceptions.

**Review by tigrannajaryan — APPROVED** [review]
LGTM.

Should we have at least one attribute that is "Required" (e.g. `graphql.operation.name` or `graphql.operation.type`)? If you know an attribute is required you can easily identify all GraphQL spans by looking for just that attibute.

It is not mandatory but it is useful to have at least one Required attribute in each set of conventions. If we don't want any to mandatory then that's ok.

**Comment by laurit** [comment]
> Should we have at least one attribute that is "Required" (e.g. `graphql.operation.name` or `graphql.operation.type`)? If you know an attribute is required you can easily identify all GraphQL spans by looking for just that attibute.
> 
> It is not mandatory but it is useful to have at least one Required attribute in each set of conventions. If we don't want any to mandatory then that's ok.

The problem here is that the input query might be invalid which makes it impossible to tell what the operation type is.

**Comment by jord1e** [comment]
> The problem here is that the input query might be invalid which makes it impossible to tell what the operation type is.

Would this not be fixed by introducing spans for parsing and validation?

This was actually included in my original GraphQL Java instrumentation PR and I was thinking of adding it back into the library: https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/4337/files#diff-6475c73d1ffe60f69dc012bbf29113111ef68d07a31fc3f15d25f6e7a904ddf1R170-R212

opentelemetry-js already has these and Apollo Gateway also has them kind of: https://www.apollographql.com/docs/federation/opentelemetry#graphql-specific-spans

We could start out with `graphql.parse`, `graphql.validate` and `graphql.execute` (in this PR `graphql` I think)

I do not know the Ruby instrumentation situation but for GraphQL Java it is definitely possible

**Comment by laurit** [comment]
> > The problem here is that the input query might be invalid which makes it impossible to tell what the operation type is.
> 
> Would this not be fixed by introducing spans for parsing and validation?
> 
I haven't read through everything in the spec, but in the parts that I have read I have not noticed a pattern where instrumentation would be required to generate multiple nested or sequential spans. Introducing `parse` and `validation` spans would deviate from this. I suspect that an attempt to introduce them would start a discussion on which sub-operations are important enough to deserve their own spans and whether such spans should be generated by default. I personally don't believe that we should have spans for `parse` and `validation` in default configuration and thus would refrain from generating them unless there are guidelines that either mandate or encourage generating such spans. I'm aware that we do have GraphQL instrumentations that generate these spans. The intent of this PR is not to disallow generating such spans, but rather to find some common ground. Hopefully the question of whether and how many spans should sub-operations produce will get answered in common guidelines for writing instrumentations.

> And I definitely agree that we should have error handling
>
Do you feel that storing GraphQL errors as exceptions isn't good enough?

**Comment by jord1e** [comment]
> This PR was marked stale due to lack of activity. It will be closed in 7 days.

No.

**Comment by tigrannajaryan** [comment]
@open-telemetry/specs-trace-approvers please review.

**Review by dyladan — APPROVED** [review]
This looks good to me. We've had multiple requests for this in JS so I'm glad to see this near the finish line.

**Comment by dyladan** [comment]
> > Should we have at least one attribute that is "Required" (e.g. `graphql.operation.name` or `graphql.operation.type`)? If you know an attribute is required you can easily identify all GraphQL spans by looking for just that attibute.
> > It is not mandatory but it is useful to have at least one Required attribute in each set of conventions. If we don't want any to mandatory then that's ok.
> 
> The problem here is that the input query might be invalid which makes it impossible to tell what the operation type is.

If the query is invalid, then possibly we should have a fallback operation name/type which is explicitly `invalid` or `failure` or similar?

**Review by jmacd — APPROVED** [review]

**Review by spencerwilson — APPROVED** [review]
Thanks for the ping, Tigran.

> I suspect that an attempt to introduce them would start a discussion on which sub-operations are important enough to deserve their own spans and whether such spans should be generated by default. ... The intent of this PR is not to disallow generating such spans, but rather to find some common ground.

I'm totally supportive of doing things iteratively, and think this PR is effective at finding a basic starting point. While I'm interested in future work that specifies semconv for particular parts of operation execution (see my issue comment: https://github.com/open-telemetry/opentelemetry-specification/issues/1670#issuecomment-997332541) —and possibly defines all those more-detailed spans to be _optional_—I think specifying those conventions is unlikely to motivate changes to what this PR defines.

Thank you, @laurit!

**Comment by andrewhr** [comment]
One concern around span names following the `<graphql.operation.type> <graphql.operation.name>` pattern, operation name can be anything the user defined, and so, they're strictly speaking unbounded. AFAICT, this goes against a hard constraint for span names, correct?

File: specification/trace/semantic_conventions/instrumentation/graphql.md [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+`graphql.operation.type` and `graphql.operation.name` are available. If `graphql.operation.name` is not available, the
+span SHOULD be named `<graphql.operation.type>`. When `<graphql.operation.type>` is not available, `GraphQL Operation`
+MAY be used as span name.
+
+<!-- semconv graphql -->
+| Attribute  | Type | Description  | Examples  | Required |
+|---|---|---|---|---|
+| `graphql.operation.name` | string | The name of the operation being executed. | `findBookById` | No |
+| `graphql.operation.type` | string | The type of the operation being executed. | `query`; `mutation`; `subscription` | No |
+| `graphql.operation.body` | string | The GraphQL query being executed. [1] | `query findBookById { bookById(id: ?) { name } }` | No |
```
Comment (andrewhr) [inline specification/trace/semantic_conventions/instrumentation/graphql.md:18]: What about `graphql.operation.document`, or maybe `graphql.document` instead? This fall in line with some of [GraphQL's terminology](https://spec.graphql.org/June2018/#sec-Language.Document).

Alternatively, `graphql.selection` as [selection](https://spec.graphql.org/June2018/#sec-Selection-Sets) applies at resolver level - the same field could be reused.

For the record, [JS implementation](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/1da0216180de694c15ec356d476f465811757ae4/plugins/node/opentelemetry-instrumentation-graphql/src/enums/AttributeNames.ts#L18) uses `graphql.source` for both purposes.

**Comment by tigrannajaryan** [comment]
@laurit please resolve the conversations and rebase from main to prepare for merging.

@open-telemetry/specs-approvers we have the minimum required number of approvals so unless I see objections I am going to merge this.

File: specification/trace/semantic_conventions/instrumentation/graphql.md (hunk 11 above)
Comment (laurit) [reply specification/trace/semantic_conventions/instrumentation/graphql.md]: Thanks for the suggestion, I believe `graphql.document` is the best name that has been suggested for this attribute this far.
Your concern about the span naming is definitely valid. Luckily these conventions are experimental and open for further refinement. Having a span name that lets users identify the operation is much more user friendly than just using operation type or `graphql.execute` as the span name.

**Comment by trask** [comment]
@laurit looks like there's a merge conflict

**Comment by laurit** [comment]
@trask thanks, rebased.

**Comment by trask** [comment]
@open-telemetry/technical-committee I think this is ready to be merged, thx!
